service/history/queues/action_slice_count.go

```diff
@@ -0,0 +1,151 @@
+)
+
+const (
+	targetThresholdCoefficient = 0.8
+)
+
+type (
+	actionSliceCount struct {
+		attributes   *AlertAttributesSlicesCount
+		monitor      Monitor
+		completionFn actionCompletionFn
+	}
+
+	compactCandidate struct {
+		slice    Slice
+		distance tasks.Key
+	}
+)
+
+func newSliceCountAction(
+	attributes *AlertAttributesSlicesCount,
+	monitor Monitor,
+	completionFn actionCompletionFn,
+) Action {
+	return &actionSliceCount{
+		attributes:   attributes,
+		monitor:      monitor,
+		completionFn: completionFn,
+	}
+}
+
+func (a *actionSliceCount) Run(readerGroup *ReaderGroup) {
+	defer a.completionFn()
+
+	// first check if the alert is still valid
+	if a.monitor.GetTotalSliceCount() <= a.attributes.CriticalSliceCount {
+		return
+	}
+
+	// then try to shrink existing slices, which may reduce slice count
+	readers := readerGroup.Readers()
+	for _, reader := range readers {
+		reader.ShrinkSlices()
+	}
+	currentSliceCount := a.monitor.GetTotalSliceCount()
+	if currentSliceCount <= a.attributes.CriticalSliceCount {
+		return
+	}
+
+	// have to compact (force merge) slices to reduce slice count
+	preferredSliceCount := int(float64(a.attributes.CriticalSliceCount) * targetThresholdCoefficient)
+	numSliceToCompact := currentSliceCount - preferredSliceCount
+
+	// find compact candidates by calculating distance between two slices
+	// and sort candidates by distance
+	candidates := make([]compactCandidate, 0, currentSliceCount)
+	for readerID, reader := range readers {
+		// first try only compacting slices in non-default reader
+		if readerID == defaultReaderId {
+			continue
+		}
+
+		candidates = a.appendCompactCandidatesForReader(candidates, reader)
+	}
+	a.sortCompactCandidates(candidates)
+
+	sliceToCompact := make(map[Slice]struct{}, numSliceToCompact)
+	for _, candidate := range candidates[:util.Min(numSliceToCompact, len(candidates))] {
+		sliceToCompact[candidate.slice] = struct{}{}
+	}
+
+	// if there are not enough slices to compact in non-default readers,
+	// have to compact slices in the default reader
+	// note here the comparision is against CriticalSliceCount, not preferredSliceCount
+	remainingToCompact := currentSliceCount - len(sliceToCompact) - a.attributes.CriticalSliceCount
+	if remainingToCompact > 0 {
+		candidates = a.appendCompactCandidatesForReader(nil, readers[defaultReaderId])
+		a.sortCompactCandidates(candidates)
+		for _, candidate := range candidates[:util.Min(remainingToCompact, len(candidates))] {
+			sliceToCompact[candidate.slice] = struct{}{}
+		}
+	}
+
+	// finally, perform the compaction
+	for _, reader := range readers {
+		reader.CompactSlices(func(s Slice) bool {
+			_, ok := sliceToCompact[s]
+			return ok
+		})
+	}
+}
+
+func (a *actionSliceCount) appendCompactCandidatesForReader(
+	candidates []compactCandidate,
+	reader Reader,
+) []compactCandidate {
+	// calculate distance between two slices
+	var prevRange *Range
+	reader.WalkSlices(func(s Slice) {
+		if prevRange == nil {
+			return
+		}
```

Comment:
when is prevRange not nil?

Reply:
It's not nil when the inline function is invoked the second time and after. Basically the idea here is to skip the first slice in the reader as there's no previous slice and can't calculate the distance.